conjure-runtime/src/service/http_error.rs

```diff
@@ -0,0 +1,418 @@
+///
+/// Propagation configuration is passed via the `PropagationConfig` struct in the request's extension map for now.
+///
+/// If `propagate_qos_errors` is `true`, 429 and 503 responses will be turned into Conjure "throttle" and "service
+/// unavailable" errors respectively. Otherwise, they run into service errors. In both cases, the error's cause will be
+/// the `ThrottledError` and `UnavailableError` types respectvely. If a `Retry-After` header is present on a 429
+/// response it will be included in the error.
+///
+/// If `propagate_service_errors` is `true`, Conjure errors returned by the server will be propagated, with the new
+/// `Error` inheriting the incoming error's code, name, instance ID, and parameters. Otherwise it will be treated as a
+/// generic internal error. In both cases, the cause will be a `RemoteError`.
+pub struct HttpErrorLayer;
+
+impl<S> Layer<S> for HttpErrorLayer {
+    type Service = HttpErrorService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        HttpErrorService { inner }
+    }
+}
+
+pub struct HttpErrorService<S> {
+    inner: S,
+}
+
+impl<S, B1, B2> Service<Request<B1>> for HttpErrorService<S>
+where
+    S: Service<Request<B1>, Response = Response<B2>, Error = Error>,
+    B2: Body,
+    B2::Error: Into<Box<dyn error::Error + Sync + Send>>,
+{
+    type Response = Response<B2>;
+    type Error = Error;
+    type Future = HttpErrorFuture<S::Future, B2>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<B1>) -> Self::Future {
+        let config = *req
+            .extensions()
+            .get::<PropagationConfig>()
+            .expect("request should contain PropagationConfig");
+
+        HttpErrorFuture::Call {
+            future: self.inner.call(req),
+            config,
+        }
+    }
+}
+
+#[pin_project(project = Projection)]
+pub enum HttpErrorFuture<F, B> {
+    Call {
+        #[pin]
+        future: F,
+        config: PropagationConfig,
+    },
+    ReadingBody {
+        status: StatusCode,
+        #[pin]
+        body: B,
+        buf: Vec<u8>,
+        config: PropagationConfig,
+    },
+}
+
+impl<F, B> Future for HttpErrorFuture<F, B>
+where
+    F: Future<Output = Result<Response<B>, Error>>,
+    B: Body,
+    B::Error: Into<Box<dyn error::Error + Sync + Send>>,
+{
+    type Output = Result<Response<B>, Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            let new_state = match self.as_mut().project() {
+                Projection::Call { future, config } => {
+                    let response = ready!(future.poll(cx))?;
+
+                    if response.status().is_success() {
+                        return Poll::Ready(Ok(response));
+                    } else if response.status() == StatusCode::TOO_MANY_REQUESTS {
```

Comment:
nit: would the braches from here dow look clearer as a match?

I was thinking this should read:

```
if response.status().is_success() {
    return Poll::Ready(Ok(respone));
}

match response.status() {
    TOO_MANY_REQUESTS => {
        // handle TOO_MANY_REQUESTS
    },
    SERVICE_UNAVAILABLE => {
        // handle SERVICE_UNAVAILABLE
    },
    _ => {
        // handle the rest
    }
}

Reply:
Done